### PR TITLE
feat(edge): EE-4570 allow pre-pull images with edge stack deployment

### PR DESCRIPTION
**Pull request by cmenginnz**

### CODE REVIEW

File: edge/poll_async.go [hunk 1]
```diff
@@ -231,24 +230,24 @@ func (service *PollService) processStackCommand(ctx context.Context, command cli
 		return newOperationError("stack", "n/a", err)
 	}
 
-	responseStatus := int(stack.EdgeStackStatusOk)
+	responseStatus := portainer.EdgeStackStatusAcknowledged
```
Comment (oscarzhou-portainer) [inline edge/poll_async.go:233]: Should this be `portainer.EdgeStackStatusOk`?

File: edge/client/portainer_edge_async_client.go [hunk 2]
```diff
@@ -409,20 +418,44 @@ func (client *PortainerAsyncClient) executeAsyncRequest(payload AsyncRequest, po
 }
 
 // SetEdgeStackStatus updates the status of an Edge stack on the Portainer server
-func (client *PortainerAsyncClient) SetEdgeStackStatus(edgeStackID, edgeStackStatus int, error string) error {
+func (client *PortainerAsyncClient) SetEdgeStackStatus(
+	edgeStackID int,
+	edgeStackStatus portainer.EdgeStackStatusType,
+	error string,
```
Comment (oscarzhou-portainer) [inline edge/client/portainer_edge_async_client.go:424]: We can change the input variable name `error` with this PR

File: edge/client/portainer_edge_async_client.go [hunk 3]
```diff
@@ -409,20 +418,44 @@ func (client *PortainerAsyncClient) executeAsyncRequest(payload AsyncRequest, po
...
+
+	details := &status.Details
+	switch edgeStackStatus {
+	case portainer.EdgeStackStatusOk:
+		details.Ok = true
+	case portainer.EdgeStackStatusError:
+		details.Error = true
+	case portainer.EdgeStackStatusAcknowledged:
+		*details = portainer.EdgeStackStatusDetails{
+			Acknowledged: true,
```
Comment (oscarzhou-portainer) [inline edge/client/portainer_edge_async_client.go:446]: Is there a reason that `Acknowledged` assignment is different from other status assignment?

File: edge/client/portainer_edge_client.go [hunk 4]
```diff
@@ -157,17 +157,23 @@ func (client *PortainerEdgeClient) GetEdgeStackConfig(edgeStackID int) (*agent.E
...
+	Status     portainer.EdgeStackStatusType
 	EndpointID portainer.EndpointID
 }
 
 // SetEdgeStackStatus updates the status of an Edge stack on the Portainer server
-func (client *PortainerEdgeClient) SetEdgeStackStatus(edgeStackID, edgeStackStatus int, error string) error {
+func (client *PortainerEdgeClient) SetEdgeStackStatus(
+	edgeStackID int,
+	edgeStackStatus portainer.EdgeStackStatusType,
+	error string,
```
Comment (oscarzhou-portainer) [inline edge/client/portainer_edge_client.go:175]: The same variable naming issue

File: edge/client/interface.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ type PortainerClient interface {
 	GetEnvironmentID() (portainer.EndpointID, error)
 	GetEnvironmentStatus(flags ...string) (*PollStatusResponse, error)
 	GetEdgeStackConfig(edgeStackID int) (*agent.EdgeStackConfig, error)
-	SetEdgeStackStatus(edgeStackID, edgeStackStatus int, error string) error
+	SetEdgeStackStatus(edgeStackID int, edgeStackStatus portainer.EdgeStackStatusType, error string) error
```
Comment (oscarzhou-portainer) [inline edge/client/interface.go:16]: variable naming

File: edge/stack/stack.go [hunk 6]
```diff
@@ -260,6 +262,57 @@ func (manager *StackManager) nextPendingStack() *edgeStack {
...
+
+func (manager *StackManager) pullImages(ctx context.Context, stack *edgeStack, stackName, stackFileLocation string) error {
+	manager.mu.Lock()
+	defer manager.mu.Unlock()
+
+	log.Debug().Int("stack_identifier", int(stack.ID)).Msg("stack pulling images")
+
+	if stack.PrePullImage || stack.RePullImage {
+		stack.Retries += 1
+		if stack.Retries <= RetryInterval || stack.Retries%RetryInterval == 0 {
```
Comment (oscarzhou-portainer) [inline edge/stack/stack.go:283]: It seems that the function will retry more than 3 times for the worst case?

**Review by oscarzhou-portainer — APPROVED** [review]
LGTM. Left some comments as below

File: edge/poll_async.go (hunk 1 above)
Comment (cmenginnz) [reply edge/poll_async.go:233]: Here the commands are enqueued and executed asynchronously. So it is ACK here.

File: edge/stack/stack.go (hunk 6 above)
Comment (cmenginnz) [reply edge/stack/stack.go:283]: In the first hour, we retry every 5s. After the first hour, we retry every 1 hour.  Retry keeps 7 days

File: edge/client/portainer_edge_async_client.go (hunk 2 above)
Comment (cmenginnz) [reply edge/client/portainer_edge_async_client.go:424]: Do you want to rename "edgeStackStatus" to "error"?

File: edge/client/portainer_edge_async_client.go (hunk 3 above)
Comment (cmenginnz) [reply edge/client/portainer_edge_async_client.go:446]: When Acknowledged is set, we reset other statuses too here.

**Pull request merged**